### PR TITLE
FractionalFactorialStrategy

### PR DESCRIPTION
**Pull request by jduerholt**
This PR adds the "FractionalFactorialStrategy", its implementation relies on `pyDOE3` and is heavily influenced by `Minitab`.

### CODE REVIEW

File: bofire/data_models/strategies/fractional_factorial.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                n_factors=len(self.domain.inputs), n_generators=self.n_generators
+            )
+        return self
+
+    @staticmethod
+    def validate_generator(n_factors: int, generator: str) -> str:
+        if len(generator.split(" ")) != n_factors:
+            raise ValueError("Generator does not match the number of factors.")
+        # clean it and transform it into a list
+        generators = [item for item in re.split(r"\-|\s|\+", generator) if item]
```
Comment (bertiqwerty) [inline bofire/data_models/strategies/fractional_factorial.py:45]: Why do you split twice and differently? One time to just to check the length? And then with the regex again?

File: bofire/data_models/strategies/fractional_factorial.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        lengthes = [len(i) for i in generators]
+
+        # Indices of single letters (main factors)
+        idx_main = [i for i, item in enumerate(lengthes) if item == 1]
+
+        # Check that single letters (main factors) are unique
+        if len([generators[i] for i in idx_main]) != len(
```
Comment (bertiqwerty) [inline bofire/data_models/strategies/fractional_factorial.py:52]:
```
if len(idx_main) != len({...}):
```

File: bofire/data_models/strategies/fractional_factorial.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            "".join(sorted([generators[i] for i in idx_main]))
+            != string.ascii_lowercase[: len(idx_main)]
+        ):
+            raise ValueError("Main factors are not in alphabetical order.")
+
+        # Indices of letter combinations (we need them to fill out H2 properly).
+        idx_combi = [i for i, item in enumerate(generators) if item != 1]
+
+        # Check that letter combinations are unique
+        if len([generators[i] for i in idx_combi]) != len(
```
Comment (bertiqwerty) [inline bofire/data_models/strategies/fractional_factorial.py:68]: see above

File: bofire/strategies/fractional_factorial.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        if self.n_repetitions > 1:
+            design = pd.concat([design] * (self.n_repetitions), ignore_index=True)
+        # setup the center points
+        centers = pd.DataFrame(
+            {key: [0] * self.n_center for key in self.domain.inputs.get_keys()}
+        )
+        design = pd.concat([design, centers], ignore_index=True)
+        # scale the design to 0 and 1
+        design = (design + 1.0) / 2.0
+        print(design)
```
Comment (bertiqwerty) [inline bofire/strategies/fractional_factorial.py:44]: debug print?

File: tests/bofire/data_models/strategies/test_fractional_factorial.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    ):
+        FractionalFactorialStrategy.get_generator(n_factors, n_generators)
+
+
+@pytest.mark.parametrize(
+    "n_factors, generator, message",
+    [
+        (2, "a b c", "Generator does not match the number of factors."),
+        (2, "a a", "Main factors are confounded with each other."),
+        (2, "a c", "Main factors are not in alphabetical order."),
```
Comment (bertiqwerty) [inline tests/bofire/data_models/strategies/test_fractional_factorial.py:56]: Actually, they are in alphabetical order. I find the error message confusing.

**Review by bertiqwerty — COMMENTED** [review]
Thanks Johannes. Maybe we should also add howto-style documentation when we add new features?

**Review by bertiqwerty — APPROVED** [review]
Please consider my comments and merge whenever you think this is ready.

**Comment by jduerholt** [comment]
> Thanks Johannes. Maybe we should also add howto-style documentation when we add new features?

You mean adding a notebook for this feature?

**Comment by bertiqwerty** [comment]
> > Thanks Johannes. Maybe we should also add howto-style documentation when we add new features?
> 
> You mean adding a notebook for this feature?

Yes. Not only for this feature, but in general. Whenever we have a feature that helps us to solve new use cases we could a short how-to. This could be a new bullet point here https://experimental-design.github.io/bofire/examples/. For instance also ENTMOOT could be presented somewhere in the docs.

File: bofire/data_models/strategies/fractional_factorial.py (hunk 1 above)
Comment (jduerholt) [reply bofire/data_models/strategies/fractional_factorial.py:45]: Yes, once for the length and to remove "Vorzeichen". THe second one is stolen directly from pyDOE3.

File: bofire/data_models/strategies/fractional_factorial.py (hunk 2 above)
Comment (jduerholt) [reply bofire/data_models/strategies/fractional_factorial.py:52]: Good catch.

File: bofire/data_models/strategies/fractional_factorial.py (hunk 3 above)
Comment (jduerholt) [reply bofire/data_models/strategies/fractional_factorial.py:68]: Thx.

File: bofire/strategies/fractional_factorial.py (hunk 4 above)
Comment (jduerholt) [reply bofire/strategies/fractional_factorial.py:44]: Indeed. Thanks.

File: tests/bofire/data_models/strategies/test_fractional_factorial.py (hunk 5 above)
Comment (jduerholt) [reply tests/bofire/data_models/strategies/test_fractional_factorial.py:56]: You are right. I changed it.

**Comment by jduerholt** [comment]
Good idea, I added an example notebook and linked it into the docs.

**Pull request merged**